<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Alex Pickering" />

<meta name="date" content="2016-04-18" />

<title>Cross-Platform Microarray Meta-Analyses</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div class="fluid-row" id="header">


<h1 class="title">Cross-Platform Microarray Meta-Analyses</h1>
<h4 class="author"><em>Alex Pickering</em></h4>
<h4 class="date"><em>2016-04-18</em></h4>

</div>


<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>This package helps in the meta-analysis of microarray data obtained from Gene Expression Omnibus (<a href="http://www.ncbi.nlm.nih.gov/gds/">GEO</a>).</p>
</div>
<div id="obtaining-raw-data" class="section level3">
<h3>Obtaining Raw Data</h3>
<p>Search GEO to find some GSEs that you want to do a meta-analysis for. For this example, I searched for LY-294002. I restricted the search to:</p>
<ul>
<li><strong>Entry type</strong>:Series</li>
<li><strong>Organism</strong>: Homo sapiens and Mus musculus</li>
<li><strong>Study type</strong>: Expression profiling by array</li>
</ul>
<p>This search produced 35 hits. The following five were chosen for this example. In practice, it is best to select all GSEs matching the necessary criteria (supplementary raw data files available, multiple samples for each treatment, and from either Affymetrix, Illumina, or Agilent mouse/human platforms).</p>
<p>After you have chosen all the GSEs that you want to run a meta-analysis on, the next step is to obtain and uncompress the raw data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(crossmeta)

<span class="co">#specify where data will be downloaded</span>
data_dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">getwd</span>(), <span class="st">&quot;data&quot;</span>, <span class="st">&quot;LY294002&quot;</span>)

<span class="co">#gather GSE names into a vector for each manufacturer</span>
affy_names  &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GSE9601&quot;</span>, <span class="st">&quot;GSE15069&quot;</span>)
illum_names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GSE50841&quot;</span>, <span class="st">&quot;GSE34817&quot;</span>, <span class="st">&quot;GSE29689&quot;</span>)

<span class="co">#obtain the raw data(supplementary files).</span>
<span class="kw">get_raw_affy</span>(affy_names, data_dir)
<span class="kw">get_raw_illum</span>(illum_names, data_dir)</code></pre></div>
<p>After the raw data has been obtained, it’s time to load it. For raw Illumina data only, you must check that it has the correct format (and edit it if needed).</p>
</div>
<div id="checking-raw-illumina-data" class="section level3">
<h3>Checking Raw Illumina Data</h3>
<p>To help, I recommend that you download and set Sublime Text 2 as your default text editor. It has very nice regular expression capabilities. <a href="https://www.cheatography.com/davechild/cheat-sheets/regular-expressions/">Here</a> is a good regular expression cheat-sheat.</p>
<p>Ensure the following:</p>
<ul>
<li>Detection p-values present(usually every second column)</li>
<li>File format is tab seperated .txt file</li>
<li>Column header formats:
<ul>
<li>probeid header is <strong>ID_REF</strong></li>
<li>expression data headers are <strong>AVG_Signal-sample_name</strong></li>
<li>pval headers are <strong>Detection-sample_name</strong></li>
</ul></li>
</ul>
<p>Here is a function to open each Illumina raw file one-by-one with your default text editor (Sublime Text 2 is recommended).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">illum_names &lt;-<span class="st"> </span><span class="kw">open_raw_illum</span>(illum_names, data_dir)</code></pre></div>
<p>For <strong>GSE50841</strong>:</p>
<ul>
<li>using Sublime Text 2, click Find –&gt; Replace and type</li>
<li><strong>Find What</strong>: <em>(SAMPLE \d+)\tDetection Pval</em></li>
<li><strong>Replace With</strong>: <em>AVG_Signal-\1\tDetection-\1</em></li>
</ul>
<p>For <strong>GSE34817</strong>:</p>
<ul>
<li>using Sublime Text 2, click Find –&gt; Replace and type</li>
<li><strong>Find What</strong>: <em>(MDA.+?)\tDetection Pval</em></li>
<li><strong>Replace With</strong>: <em>AVG_Signal-\1\tDetection-\1</em></li>
</ul>
<p>For <strong>GSE29689</strong> just change PROBE_ID to ID_REF.</p>
</div>
<div id="loading-data" class="section level3">
<h3>Loading Data</h3>
<p>Now that the data is downloaded (and Illumina data is in correct format), we must load and annotate the data. Annotation requires the ~4Gb GEOmetadb.sqlite file in your working directory (to lookup bioconductor annotation data packages). Transfer it here if you have it, otherwise it will be downloaded. All relevant bioconductor annotation data packages will also be downloaded.</p>
<p>If GEOmetadb.sqlite lacks information on the bioconductor annotation data package for a given platform, entry will be requested. Use the default entry (platform title) to search online for the correct package name.</p>
<p>For GSE29689, my version of GEOmetadb.sqlite lacked the bioconductor annotation data package. A popup with a default value of “Illumina HumanRef-8 v3.0 expression beadchip” came up, so I went to <a href="https://bioconductor.org/">Bioconductor</a> and searched “illuminahuman” to find the appropriate package (illuminaHumanv3).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">affy_esets &lt;-<span class="st"> </span><span class="kw">load_affy</span>(affy_names, data_dir)
illum_esets &lt;-<span class="st"> </span><span class="kw">load_illum</span>(illum_names, data_dir)

<span class="co">#put esets together</span>
esets &lt;-<span class="st"> </span><span class="kw">c</span>(affy_esets, illum_esets)</code></pre></div>
<p>The next step is to select only features with gene symbols that are common between all platforms.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">com_esets &lt;-<span class="st"> </span><span class="kw">commonize</span>(esets)</code></pre></div>
</div>
<div id="differential-expression" class="section level3">
<h3>Differential Expression</h3>
<p>We are now ready for differential expression analysis! Select control then test samples for each contrast you want to analyse. Once done for a study, hit “Cancel” and you will be prompted to do the same for the next study. If you re-use a control group for multiple contrasts in a given study, use the same combination of tissue and group name (because it’s the same group!). Analogously, do not give different groups the same name for contrasts within a given study.</p>
<p>It’s useful to have a <a href="http://www.ncbi.nlm.nih.gov/geo/">GEO</a> tab open in your browser to lookup experimental details (e.g. tissue and group) for GSEs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">anals &lt;-<span class="st"> </span><span class="kw">diff_expr</span>(com_esets, data_dir)</code></pre></div>
<p>If you realize that you selected the wrong groups or that you need to add or remove a GSE (and have to run <code>commonize</code> again with the updated eset list), you can avoid having to reselect samples and rename groups that you already selected. For example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#load auto-saved results of diff_expr that you previously selected samples for</span>
prev_names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GSE9601&quot;</span>, <span class="st">&quot;GSE15069&quot;</span>)
prev &lt;-<span class="st"> </span><span class="kw">load_diff</span>(prev_names, data_dir)

<span class="co">#supply prev to diff_expr</span>
anals &lt;-<span class="st"> </span><span class="kw">diff_expr</span>(com_esets, data_dir, <span class="dt">prev_anals=</span>prev)</code></pre></div>
<p>Multi-dimensional scaling plots are generated for each contrast. These plots are generated from data with the effects of surrogate variables removed.</p>
</div>
<div id="meta-analysis" class="section level3">
<h3>Meta-Analysis</h3>
<p>Now that differential expression analyses have been performed for each study, let’s combine the studies.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#re-load previous analyses if need to</span>
gse_names  &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GSE9601&quot;</span>, <span class="st">&quot;GSE15069&quot;</span>, <span class="st">&quot;GSE50841&quot;</span>, <span class="st">&quot;GSE34817&quot;</span>, <span class="st">&quot;GSE29689&quot;</span>)
anals &lt;-<span class="st"> </span><span class="kw">load_diff</span>(gse_names, data_dir)

<span class="co">#MetaArray object (effect of SVs removed)</span>
ma_sva &lt;-<span class="st"> </span><span class="kw">make_ma</span>(anals, <span class="dt">sva=</span>T)</code></pre></div>
<p>Now the whole world of <a href="http://is.muni.cz/www/184415/MAMA_full.pdf">MAMA</a> meta-analyses methods are available to you.</p>
<p>My favorite (which happens to integrate with my ccmap package - comming soon) is shown:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(MAMA)

es &lt;-<span class="st"> </span><span class="kw">ES.GeneMeta</span>(ma_sva, <span class="st">&quot;treatment&quot;</span>, <span class="dt">nperm=</span><span class="dv">100</span>)</code></pre></div>
<p>Let’s save up then go find a drug combination to reverse or mimic your signature!</p>
<p>ccmap (<strong>C</strong>ombination <strong>C</strong>onnectivity <strong>M</strong>aping comming soon)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">saveRDS</span>(es, <span class="st">&quot;LY_es.rds&quot;</span>)</code></pre></div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
